### PR TITLE
add `D.Cauchy`

### PR DESCRIPTION
**Pull request by YoshikawaMasashi**
split from #4678.

### CODE REVIEW

**Comment by YoshikawaMasashi** [comment]
Jenkins, test this please.

File: chainer/distributions/cauchy.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    def _is_gpu(self):
+        return isinstance(self.loc.data, cuda.ndarray)
+
+    def log_prob(self, x):
+        return - numpy.log(numpy.pi) + exponential.log(self.scale) \
+            - exponential.log((x - self.loc)**2 + self.scale**2)
+
+    @property
+    def mean(self):
+        return self.loc * numpy.nan
```
Comment (toslunar) [inline chainer/distributions/cauchy.py]: Could you add a warning on that a mean of a Cauchy distribution is (mathematically) undefined?

Could you explain why the output depends (in the graph) on the variable `self.loc`?

Comment (YoshikawaMasashi) [reply chainer/distributions/cauchy.py]: > Could you explain why the output depends (in the graph) on the variable self.loc?
sorry, it is needless. I'll fix.

**Comment by YoshikawaMasashi** [comment]
Jenkins, test this please.

File: tests/chainer_tests/distributions_tests/test_cauchy.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    def test_sample_cpu(self):
+        self.check_sample(False)
+
+    @attr.gpu
+    def test_sample_gpu(self):
+        self.check_sample(True)
+
+    def check_variance(self, is_gpu):
+        with self.assertWarns(RuntimeWarning):
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_cauchy.py]: ditto

File: chainer/distributions/cauchy.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    @property
+    def support(self):
+        return 'real'
+
+    @property
+    def variance(self):
+        warnings.warn("Variance of the cauchy distribution is undefined.",
+                      RuntimeWarning)
+        xp = cuda.get_array_module(self.loc)
+        return chainer.as_variable(xp.full(self.loc.shape, xp.nan))
```
Comment (toslunar) [inline chainer/distributions/cauchy.py]: ditto

File: chainer/distributions/cauchy.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    def backward(self, target_input_indexes, grad_outputs):
+        gy, = grad_outputs
+        x, = self.get_retained_inputs()
+        return gy / trigonometric.cos(
+            utils.force_array(self.h, x.dtype)) ** 2 * numpy.pi,
+
+
+def _cauchy_icdf(x):
+    y, = CauchyICDF().apply((x,))
```
Comment (toslunar) [inline chainer/distributions/cauchy.py]: Does the following work?
```
    h = (x - 0.5) * numpy.pi
    y = chainer.functions.tan(h)
```

**Comment by YoshikawaMasashi** [comment]
Jenkins, test this please.

File: tests/chainer_tests/distributions_tests/test_cauchy.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
+import numpy
+import unittest
```
Comment (toslunar) [inline tests/chainer_tests/distributions_tests/test_cauchy.py]:
```
import unittest

import numpy

...
```

File: chainer/distributions/cauchy.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+import numpy
+import warnings
```
Comment (toslunar) [inline chainer/distributions/cauchy.py]:
```
import warnings

import numpy

...
```

**Pull request merged**